src/couch/src/couch_time_seq.erl

```diff
@@ -0,0 +1,312 @@
+
+-export([
+    new/0,
+    new/1,
+    check/1,
+    since/2,
+    update/2,
+    update/3,
+    histogram/2
+]).
+
+-export_type([
+    time_seq/0
+]).
+
+% How bin count = 60 was picked:
+%
+%  - With the ?INTERVALS schedule defined below ran 1 update per hour for 1M
+%    updates starting at year 3000 and ending at year 3114 and obtained:
+%      * Less than 10 years of spacing between years at the start: 3000, 3009, 3018 ...
+%      * Ten individual latest days: 3114-01-20 -> 3114-01-30
+%      * Seven individual latest months: 3113-07 -> 3114-01
+%  - Uncompressed term_to_binary(TSeq) = 920B
+%  - RAM flat size erts_debug:flat_size(TSeq) * erlang:system_info(wordsize) = 2KB
+%
+-define(MAX_BIN_COUNT, 60).
+
+-define(H, 3600).
+-define(D, ?H * 24).
+-define(M, ?D * 30).
+-define(Y, ?D * 365).
+%% erlfmt-ignore
+-define(INTERVALS, [
+    ?H * 3, ?H * 6, ?H * 12,
+    ?D, ?D * 10,
+    ?M, ?M * 6,
+    ?Y, ?Y * 2, ?Y * 4, ?Y * 8, ?Y * 16
+]).
+
+% Version number so we could upgrade the data structure in the future.
+%
+-define(VER, 1).
+
+% Users can set a minimum time boundary to avoid errors with broken clocks.
+% Sometimes embedded systems get booted into 1970 and then get their time from
+% NTP. Updates are ignored if we can tell they are obviously wrong. Use some
+% recent time as a default min time value:
+%
+%   1752724800 = 2025-08-01T00:00:00Z
+%
+-define(DEFAULT_MIN_TIME, 1754006400).
+
+-type unix_time() :: pos_integer().
+-type bin_size() :: pos_integer().
+-type update_seq() :: non_neg_integer() | now.
+-type bin() :: {unix_time(), bin_size(), update_seq()}.
+-type time_seq() :: #{v := pos_integer(), bins := [bin()]}.
+
+-spec new() -> time_seq().
+new() ->
+    #{v => ?VER, bins => []}.
+
+-spec new(time_seq()) -> time_seq().
+new(#{v := ?VER, bins := Bins} = Ctx) when is_list(Bins) ->
+    % Future upgrade clauses would go here. When upgrading make sure to add a
+    % clause to check/1 to return true for the older version as well. But the
+    % upgarde itself will happen in new/1.
+    Ctx.
+
+-spec check(time_seq()) -> boolean().
+check(#{v := ?VER, bins := Bins}) when is_list(Bins) ->
+    % On version upgrades add a clause to return true for the old version here
+    % as well. The upgrade will happen in new/1, but the check/1 function
+    % should accept old versions as well (return true).
+    true;
+check(_) ->
+    false.
+
+-spec update(time_seq(), update_seq()) -> time_seq().
+update(#{v := ?VER} = Ctx, Seq) when is_integer(Seq), Seq >= 0 ->
+    update(Ctx, now_unix_sec(), Seq).
+
+-spec update(time_seq(), unix_time(), update_seq()) -> time_seq().
+update(#{v := ?VER} = Ctx, Time, Seq) when is_integer(Time), is_integer(Seq), Seq >= 0 ->
+    #{bins := Bins} = Ctx,
+    RoundTime = round_time(Time),
+    case Time >= min_time() of
+        true -> Ctx#{bins := update_bins(Bins, RoundTime, Seq)};
+        false -> Ctx
+    end.
+
+% Return a highest known sequence that comes before the given time. If the time
```

Comment:
Not really a problem, just a thought.

The erlang-lsp has understanding of edoc format. There is also eldoc-box which is for eglot. This means that if function documentation is written in edoc instead of a plain comment the user would get a nice behaviour. Namely a pop up (in vscode) or documentation in the echo area (emacs) when you mouse hover on function invocation. Since you are adding the documentation comment anyway. You could wrap it in `@doc`/`@end` pair. 

```
% @edoc Return a highest known sequence that comes before the given time. If the time
% falls before the oldest bin then return 0. If the time is before the time of
% the first bin plus the size of size of the bin, return atom 'now'.
% @end
```

Reply:
Add the tags for the function docs. Thanks for the suggestion.